pkg/experiment/ingester/service.go

```diff
@@ -0,0 +1,237 @@
+		logger:        log,
+		reg:           reg,
+		dbConfig:      dbConfig,
+		storageBucket: storageBucket,
+		//limits:        limits,
+	}
+
+	var (
+		err error
+	)
+
+	i.lifecycler, err = ring.NewLifecycler(
+		cfg.LifecyclerConfig,
+		&ingesterFlusherCompat{i},
+		"segment-writer-ring-name",
+		"segment-writer-ring-key",
+		true,
+		i.logger, prometheus.WrapRegistererWithPrefix("pyroscope_segment_writer_", i.reg))
+	if err != nil {
+		return nil, err
+	}
+
+	i.subservices, err = services.NewManager(i.lifecycler)
+	if err != nil {
+		return nil, errors.Wrap(err, "services manager")
+	}
+	//i.limiters = newLimiters(i.limitsForTenant)
+	if storageBucket == nil {
+		return nil, errors.New("storage bucket is required for segment writer")
+	}
+	if metastoreClient == nil {
+		return nil, errors.New("metastore client is required for segment writer")
+	}
+	metrics := newSegmentMetrics(i.reg)
+
+	i.segmentWriter = newSegmentWriter(i.phlarectx, i.logger, metrics, i.dbConfig, storageBucket, cfg.SegmentDuration, metastoreClient)
+	i.subservicesWatcher = services.NewFailureWatcher()
+	i.subservicesWatcher.WatchManager(i.subservices)
+	i.Service = services.NewBasicService(i.starting, i.running, i.stopping)
+	return i, nil
+}
+
+func (i *SegmentWriterService) starting(ctx context.Context) error {
+	return services.StartManagerAndAwaitHealthy(ctx, i.subservices)
+}
+
+func (i *SegmentWriterService) running(ctx context.Context) error {
+	select {
+	case <-ctx.Done():
+		return nil
+	case err := <-i.subservicesWatcher.Chan(): // handle lifecycler errors
+		return fmt.Errorf("lifecycler failed: %w", err)
+	}
+}
+
+func (i *SegmentWriterService) stopping(_ error) error {
+	errs := multierror.New()
+	errs.Add(services.StopManagerAndAwaitStopped(context.Background(), i.subservices))
+	errs.Add(i.segmentWriter.Stop())
+	return errs.Err()
+}
+
+func (i *SegmentWriterService) Push(ctx context.Context, req *connect.Request[segmentWriterV1.PushRequest]) (*connect.Response[segmentWriterV1.PushResponse], error) {
+	tenantID, err := tenant.ExtractTenantIDFromContext(ctx)
+	if err != nil {
+		return nil, connect.NewError(connect.CodeInvalidArgument, err)
+	}
+	var waits = make(map[segmentWaitFlushed]struct{}, len(req.Msg.Series))
+	for _, series := range req.Msg.Series {
+		var shard = shardKey(series.Shard)
+		wait, err := i.segmentWriter.ingest(shard, func(segment segmentIngest) error {
+			return i.ingestToSegment(ctx, segment, series, tenantID)
+		})
+		if err != nil {
+			return nil, err
+		}
+		waits[wait] = struct{}{}
+	}
+	if i.cfg.Async {
+		return connect.NewResponse(&segmentWriterV1.PushResponse{}), nil
+	}
+	t1 := time.Now()
+	for wait := range waits {
+		if err = wait.waitFlushed(ctx); err != nil {
+			i.segmentWriter.metrics.segmentFlushTimeouts.WithLabelValues(tenantID).Inc()
+			i.segmentWriter.metrics.segmentFlushWaitDuration.WithLabelValues(tenantID).Observe(time.Since(t1).Seconds())
+			level.Error(i.logger).Log("msg", "flush timeout", "err", err)
```

Comment:
We assume that the error indicates a timeout. We probably want to check the error type here (or `context.Err()`)